Adhoc-filtering for prometheus dashboards

davkal opened the pull request:
- Implements adhoc filters for prometheus panels (see magnifying icons
  in result tables)
- fixes templatesrv adhoc vars for panels with default datasource (datasource is
  null) which breaks when iterating over variables origininating from
  such panels
- uses `addLabelToQuery` to apply filter to query

Fixes #12908 (by actually implementing filtering)

Supersedes: #12920 (tries to hide icons, should be closed when this is accepted)

Related: #6138, but is much simpler (leaving out tag suggestions). Another draw-back is that is uses the globally shared operators which include the non-applicable `>` and `<`.

davkal committed:
Adhoc-filtering for prometheus dashboards

- Implements adhoc filters for prometheus panels (see magnifying icons
  in result tables)
- fixes templatesrv adhoc vars for panels with default datasource (datasource is
  null) which breaks when iterating over variables origininating from
  such panels
- adhoc filters now probe if the datasource supports tags
- renamed value placeholder in adhoc filters from 'select tag value' to
  'select value'
- uses `addLabelToQuery` to apply filter to query

torkelo commented:
Nice. 
since getTagKeys getTagValues are not implemented won't this result in errors? A unit test scenario where getAdHocFiters returns 1 or 2 filters would be nice.

marefr commented:
There's no way in current implementation to remove ad-hoc filters. Possibly related to error happen since getTagKeys and getTagValues is lacking.

![image](https://user-images.githubusercontent.com/1668778/45298538-16ffdc00-b509-11e8-99dd-c8c2f7cb3bee.png)

Seems like adding multiple ad-hoc filters only use one of them in the actual query:

![image](https://user-images.githubusercontent.com/1668778/45298561-297a1580-b509-11e8-8036-3a0cdf6fe47c.png)

davkal commented:
> since getTagKeys getTagValues are not implemented won't this result in errors?

Already pushed that probing.

> A unit test scenario where getAdHocFiters returns 1 or 2 filters would be nice.

I'll give it a go.

marefr commented:
Can't still remove an adhoc filter and my latest comment still applies

marefr commented:
1. Had a dashboard with a table panel using prometheus datasource. 
2. Added two adhoc filters from table
3. Then went into settings
4. Then into variables. 
5. Looks like url is constantly changing just by looking at it. 
6. Closing settings 
7. Url is constantly changing and one query is executed against prometheus kind of each second or faster and never stops.

Seems like url first is like this before entering settings/variables:
http://localhost:3000/d/YAcO1Momk/prom-test?orgId=1&var-Filters=instance%7C=%7C127.0.0.1:9091&var-Filters=job%7C=%7Cfake-data-gen

But then after it becomes like this (with array notation):
http://localhost:3000/d/YAcO1Momk/prom-test?editview=templating&orgId=1&var-Filters[]=instance%7C=%7C127.0.0.1:9091&var-Filters[]=job%7C=%7Cfake-data-gen&panelId=&fullscreen=&edit=

davkal committed:
Fixed selector parsing and extracted addLabelToQuery

davkal commented:
- Fixed multi-label support for adhoc filters
- extracted `addLabelToQuery`
- added tests

Still don't know why I cannot remove filters. @marefr any idea?

davkal commented:
> Looks like url is constantly changing just by looking at it.

@marefr I've seen this once before. Good repro steps. Seems to happen independent of adhoc filters. I'll open an issue.

davkal committed:
Return empty tag lists for adhoc filters where the ds has no tags

marefr reviewed (approved):
LGTM :+1:

bobrik commented:
This mentioned in 5.3.0 betas, but I don't see it in final release or on play.grafana.org:

![image](https://user-images.githubusercontent.com/89186/46826895-0fbc3000-cd4c-11e8-8b83-d82da6a4952c.png)

What's the status of this?

marefr commented:
@bobrik Should probably been a better description of this feature. It's not full ad-hoc query support for Prometheus. What you can do if you have a table panel is hover a cell and click on the plus/minus to active the ad-hoc filter capabilities. Try here for example: https://play.grafana.org/d/000000065/prometheus-console-tables?orgId=1
